mapper/orm/src/main/java/org/hibernate/search/mapper/orm/batchindexing/impl/IdentifierConsumerDocumentProducer.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.orm.batchindexing.impl;
+
+import java.io.Serializable;
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import javax.transaction.TransactionManager;
+
+import org.hibernate.CacheMode;
+import org.hibernate.Criteria;
+import org.hibernate.FlushMode;
+import org.hibernate.LockMode;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.criterion.CriteriaSpecification;
+import org.hibernate.criterion.Restrictions;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.engine.transaction.jta.platform.spi.JtaPlatform;
+import org.hibernate.internal.CriteriaImpl;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.pojo.work.spi.PojoWorkPlan;
+import org.hibernate.search.util.impl.common.LoggerFactory;
+
+/**
+ * This {@code SessionAwareRunnable} is consuming entity identifiers and
+ * producing corresponding {@code AddLuceneWork} instances being forwarded
+ * to the index writing backend.
+ * It will finish when the queue it is consuming from will
+ * signal there are no more identifiers.
+ *
+ * @author Sanne Grinovero
+ */
+public class IdentifierConsumerDocumentProducer implements Runnable {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final ProducerConsumerQueue<List<Serializable>> source;
+	private final SessionFactory sessionFactory;
+	private final CacheMode cacheMode;
+	private final Class<?> type;
+	private final String idName;
+	private final CountDownLatch producerEndSignal;
+	private final Integer transactionTimeout;
+	private final String tenantId;
+	private final PojoWorkPlan workPlan;
+
+	/**
+	 * The JTA transaction manager or {@code null} if not in a JTA environment
+	 */
+	private final TransactionManager transactionManager;
+
+	public IdentifierConsumerDocumentProducer(
+			ProducerConsumerQueue<List<Serializable>> fromIdentifierListToEntities,
+			SessionFactory sessionFactory,
+			CountDownLatch producerEndSignal,
+			CacheMode cacheMode, Class<?> indexedType, String idName,
+			Integer transactionTimeout,
+			String tenantId, PojoWorkPlan workPlan) {
+		this.source = fromIdentifierListToEntities;
+		this.sessionFactory = sessionFactory;
+		this.cacheMode = cacheMode;
+		this.type = indexedType;
+		this.idName = idName;
+		this.producerEndSignal = producerEndSignal;
+		this.transactionTimeout = transactionTimeout;
+		this.tenantId = tenantId;
+		this.workPlan = workPlan;
```

Comment:
As I said somewhere else, there shoudn't be a work plan here: instances of this class is supposed to be used concurrently, and work plan is not thread safe. The work plan should be created in `run()` and passed to other methods.